### PR TITLE
[CUDA][64-bit indexing] Support 64-bit indexing in `distribution_elementwise_grid_stride_kernel`

### PR DESCRIPTION
**Pull request by eqy**
For #141544 
Overhead doesn't seem to be noticeable even on small sizes (e.g., 2**10 elements)

cc @fritzo @neerajprad @alicanb @nikitaved

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Comment by ngimel** [comment]
In my example numel was 2**31-2 which should fit in int, and I don't see where it would overflow. Do you know why it was failing?

**Comment by eqy** [comment]
e.g.,
```  
int rounded_size = ((numel - 1)/(blockDim.x * gridDim.x * unroll_factor)+1) *
      blockDim.x * gridDim.x * unroll_factor;
```
looked pretty suspicious in the kernel

**Comment by ngimel** [comment]
Ah right, that would round it to 2**31 or so, depending on blockDim * gridDim. Thanks for the fix!

**Review by ngimel — APPROVED** [review]

File: test/test_cuda.py [hunk 1]
```diff
@@ -1051,6 +1052,12 @@ def run(dev: torch.device) -> int:
             abs(run(torch.device("cuda")) - run(torch.device("cpu"))) < 10_000
         )
 
+    @largeTensorTest("20GB", "cuda")
+    def test_randint_generation_for_large_numel(self) -> None:
+        numel = 2**31 + 1
+        s = torch.randint(2, (numel,), device="cuda", dtype=torch.int8).sum()
```
Comment (ngimel) [inline test/test_cuda.py:1058]: I think if you don't do `sum` here you'd be able to get away with less memory, `sum` upcasts to int64. If we are just doing a smoke test, checking `sum` on the first 1K elements should be enough, the probability that it's 0 is 2^(-1000) = 10^(-300), no one should be that unlucky.

**Pull request closed without merging**